<html>
<head>
<title>Ensj Jython</title>
</head>

<body>

<h2>Jython</h2>

<p>It is possible to write Python programs that use the ensj library
to access Ensembl databases. This is achieved by running the program
through the Jython interpreter.
</p>

<p>
Python programs can import the ensj classes directly and create their
own drivers for later use.
</p>

<p>
An alternative is to import <a
href="src_jython/ensembl.py.html">ensembl.py</a> <a href="src_jython/ensembl.py">(source)</a>, (<a
href="download.html">latest
versions</a>)
which provides a facade for ensj and automatically loads drivers from
predefined config files. To use ensembl.py add <code>from ensembl
import *</code> to your program and ensure the module is available on
your path.
</p>

<p>Here is a simple example that fetches all the genes from a genomic
location and prints their names and how many exons they have. Note
that in addition to importing <code>ensembl</code> we also need to
qualify the driver, we use <code>ensembl.human</code> rather than
<code>human</code>. It is also possible to do <code> from ensembl
import human ... human.ga.fetch(...)</code> if you don't want to use
the <code>ensembl</code> prefix.
</p>

<pre><code>
#!/usr/bin/env jython

import ensembl

# Using the current human data release
# find the number of exons for all of the genes in the region 20m 
# to 21m bases on chromosome 22.

genes = ensembl.human.ga.fetch(ensembl.Location("chromosome:22:20m-21m"))
for gene in genes:
    print gene.accessionID, gene.exons.size()
</code></pre>

<p>You can make the script executable on Unix by doing chmod +x
myscript.py. Alternatively pass it the jython interpreter
directly <code>jython myscript.py</code>

<p>Here are some other examples:</p>

<dl>
<dt><a href="src_jython/examples/all_genes.py.html">all_genes.py</a> <a href="src_jython/examples/all_genes.py">(source)</a>
 <dd>lists all human genes and their descriptions
<dt><a
 href="src_jython/examples/ensdescribe.py.html">ensdescribe.py</a> <a href="src_jython/examples/ensdescribe.py">(source)</a>
 <dd>adds descriptions and gene names to a tab-delimited text file keyed on gene accession ID
<dt><a href="src_jython/examples/enspeptidize.py.html">enspeptidize.py</a> <a href="src_jython/examples/enspeptidize.py">(source)</a>
 <dd>adds peptide accession IDs to a tab-delimited text file keyed on gene accession ID
</dl>


<p>
<b>What is Jython?</b> <a
href="http://www.jython.org/">Jython</a> is a java implementation of
the Python interpreter. It converts python scripts into java bytecode
for execution on a JVM. It supports python language constructs and
standard modules in addition to allowing access to java types. This
means we can write python programs that use ensj's java classes.
</p>

</body>
</html>
